articles/intraprocess_communication.md

```diff
@@ -0,0 +1,437 @@
+
+Moreover, the meta-messages could be delivered also to nodes in different processes if they have intra-process communication enabled. As before, the messages would be discarded immediately after being received, but they would still affect the
+performances. The overhead caused by the additional publication of meta-messages can be potentially reduced by appending to the intra-process topic names a process specific identifier.
+
+## Proposed implementation
+
+### Overview
+
+The new proposal for intra-process communication addresses all the issues previously mentioned. It has been designed with performance in mind, so it avoids any communication through the middleware between nodes in the same process. Moreover, it
+supports all the ROS2 Quality of Service configurations.
+
+Consider a simple scenario, consisting of `Publisher`s and `Subscription`s all in the same process and with the durability QoS set to `volatile`. The proposed implementation creates one buffer per `Subscription`. When a message is published to a
+topic, its `Publisher` pushes the message into the buffer of each of the `Subscription`s related to that topic and raises a notification, waking up the executor. The executor can then pop the message from the buffer and trigger the callback of the
+`Subscription`.
+
+![Proposed IPC Block Diagram](../img/intraprocess_communication/new_ipc.png)
+
+The choice of having independent buffers for each `Subscription` leads to the following advantages:
+ - It is easy to support different QoS for each `Subscription`, while, at the same time, simplifying the implementation.
+ - Multiple `Subscription`s can extract messages from their own buffer in parallel without blocking each other, thus providing an higher throughput.
+
+The only drawback is that the system is not reusing as much resources as possible, compared to sharing buffers between entities. However, from a practical point of view, the memory overhead caused by the proposed implementation with respect to the
+current one, will always be only a tiny delta compared to the overall memory usage of the application.
+
+There are three possible data-types that can be stored in the buffer:
+
+- `MessageT`
+- `shared_ptr<const MessageT>`
+- `unique_ptr<MessageT>`
+
+The choice of the buffer data-type is controlled through an additional field in the `SubscriptionOptions`. The default value for this option is denominated `CallbackDefault`, which corresponds to selecting the type between `shared_ptr<constMessageT>`
+and `unique_ptr<MessageT>` that better fits with its callback type. This is deduced looking at the output of `AnySubscriptionCallback::use_take_shared_method()`.
+
+If the history QoS is set to `keep all`, the buffers are dynamically allocated. On the other hand, if the history QoS is set to `keep last`, the buffers have a size equal to the depth of the history and they act as ring buffers (overwriting the
```

Comment:
This is different from how these QoS settings work in DDS. Not having the same semantics for QoS settings between inter-process and intra-process would make it hard to reuse a node.

Reply:
We meant to say "buffers are dynamically adjusted in size _up to some max limit_ specified as part of KEEP_ALL QoS". Then the semantics are the same, right?